src/driver/virtio.c

```diff
@@ -0,0 +1,411 @@
+	if (read8(dev, VIRTIO_PCI_STATUS) == VIRTIO_CONFIG_STATUS_FAILED)
+		error("Device signaled unrecoverable error");
+}
+
+static void virtio_legacy_setup_tx_queue(struct ixy_device* dev, uint16_t idx) {
+	if (idx != 1 && idx != 2)
+		error("Can't setup queue %u as Tx queue", idx);
+
+	// Create virt queue itself - Section 4.1.5.1.3
+	write16(dev, idx, VIRTIO_PCI_QUEUE_SEL);
+	uint32_t max_queue_size = read32(dev, VIRTIO_PCI_QUEUE_NUM);
+	info("Max queue size of tx queue #%u: %u", idx, max_queue_size);
+	if (max_queue_size == 0)
+		return;
+	size_t virt_queue_mem_size = vring_size(max_queue_size, 4096);
+	struct dma_memory mem = memory_allocate_dma(virt_queue_mem_size, true);
+	memset(mem.virt, 0xab, virt_queue_mem_size);
+	info("Allocated %zu bytes for virt queue at %p", virt_queue_mem_size, mem.virt);
+	write32(dev, mem.phy >> VIRTIO_PCI_QUEUE_ADDR_SHIFT, VIRTIO_PCI_QUEUE_PFN);
+
+	// Section 2.4.2 for layout
+	struct virtqueue* vq = calloc(1, sizeof(*vq) + sizeof(void*) * max_queue_size);
+	vring_init(&vq->vring, max_queue_size, mem.virt, 4096);
+	info("vring desc: %p, vring avail: %p, vring used: %p", vq->vring.desc, vq->vring.avail, vq->vring.used);
+	for (size_t i = 0; i < vq->vring.num; ++i) {
+		vq->vring.desc[i].len = 0;
+		vq->vring.desc[i].addr = 0;
+		vq->vring.desc[i].flags = 0;
+		vq->vring.desc[i].next = 0;
+		vq->vring.avail->ring[i] = 0;
+		vq->vring.used->ring[i].id = 0;
+		vq->vring.used->ring[i].len = 0;
+	}
+	vq->vring.used->idx = 0;
+	vq->vring.avail->idx = 0;
+	vq->vq_used_last_idx = 0;
+
+	// Section 4.1.4.4
+	uint32_t notify_offset = read16(dev, VIRTIO_PCI_QUEUE_NOTIFY);
+	info("vq notifcation offset %u", notify_offset);
+	vq->notification_offset = notify_offset;
+
+	// Ctrl queue packets are not supplied by the user
+	if (idx == 2)
+		vq->mempool = memory_allocate_mempool(max_queue_size, 2048);
+
+	// Disable interrupts - Section 2.4.7
+	vq->vring.avail->flags = VRING_AVAIL_F_NO_INTERRUPT;
+	vq->vring.used->flags = 0;
+
+	if (idx == 1)
+		dev->tx_queues = vq;
+	else
+		dev->ctrl_queues = vq;
+}
+
+static void virtio_legacy_send_command(struct ixy_device* dev, void* cmd, size_t cmd_len) {
+	struct virtqueue* vq = dev->ctrl_queues;
+
+	if (cmd_len < sizeof(struct virtio_net_ctrl_hdr))
+		error("Command can not be shorter than control header");
+	if (((uint8_t*)cmd)[0] != VIRTIO_NET_CTRL_RX)
+		error("Command class is not supported");
+
+	rte_mb();
+	// Find free desciptor slot
+	uint16_t idx = 0;
+	for (idx = 0; idx < vq->vring.num; ++idx) {
+		struct vring_desc* desc = &vq->vring.desc[idx];
+		if (desc->addr == 0)
+			break;
+	}
+	if (idx == vq->vring.num)
+		error("command queue full");
+	else
+		info("Found free desc slot at %u (%u)", idx, vq->vring.num);
+
+	struct pkt_buf* buf = pkt_buf_alloc(vq->mempool);
+	if (!buf)
+		error("Control queue ran out of buffers");
+	memcpy(buf->data, cmd, cmd_len);
+	vq->virtual_addresses[idx] = buf;
+
+#if 0 // This kills the VM, but should be allowed with VIRTIO_F_ANY_LAYOUT
```

Comment:
remove #if, but you can keep commented out code with appropriate comment

qemu error was:

kvm: virtio-net ctrl missing headers

QEMU emulator version 2.7.1 pve-qemu-kvm_2.7.1-4, Copyright (c) 2003-2016 Fabrice Bellard and the QEMU Project developers